### PR TITLE
feat(node): Upgrade to node 18 (hydrogen)

### PR DESCRIPTION
**Pull request by greg-in-a-box**
BREAKING CHANGE: Upgrade to node 18 (hydrogen)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -119,7 +119,7 @@
         }
     },
     "engines": {
-        "node": ">=14.0.0 <15.0.0",
+        "node": ">=18.16.0 <19.0.0",
```
Comment (tjuanitas) [inline package.json:122]: should we just remove the upper restriction and do something like?
`"node": ">=14.x",` or `"node": ">=16.x",`

this way it doesn't force apps to immediately upgrade and parent apps can try to use v20 if they want

**Review by tjuanitas — COMMENTED** [review]
we'd need to include breaking change in the title

File: package.json (hunk 1 above)
Comment (jpan-box) [reply package.json:122]: how would this work if the use of that legacy openssl flag is only available in 17 and above? would parent apps still be required to immediately upgrade?

the openssl flag is only invoked in local dev or build using webpack, but i dont know what impact that has on parent apps.

Comment (tjuanitas) [reply package.json:122]: depends on if versions lower than node v17 will ignore flags they don't recognize

Comment (jpan-box) [reply package.json:122]: BUIE certainly doesn't build when the flag is included on node 14, but in a parent app, they're not building buie right? or do they, when the whole parent app is built?

File: package.json [hunk 2]
```diff
@@ -42,18 +42,18 @@
...
         "build:dev:es": "BABEL_ENV=development NODE_ENV=development yarn build:es --copy-files --source-maps inline --watch --ignore \"**/*.d.ts,**/__tests__/**,**/__mocks__/**\"",
         "build:es": "babel src --extensions '.js,.tsx,.ts' --out-dir es",
         "build:i18n": "props2es",
         "build:npm": "npm-run-all clean build:i18n build:prod:npm build:prod:es",
         "build:prod:analyze": "BUNDLE_ANALYSIS=true npm-run-all setup build:prod:npm",
         "build:prod:es": "BABEL_ENV=npm NODE_ENV=production yarn build:es --source-maps --ignore \"**/*.d.ts,**/__tests__/**,**/__mocks__/**\"",
-        "build:prod:dist": "NODE_ENV=production webpack --config scripts/webpack.config.js --mode production",
+        "build:prod:dist": "NODE_ENV=production NODE_OPTIONS=--openssl-legacy-provider webpack --config scripts/webpack.config.js --mode production",
         "build:prod:examples": "LANGUAGE=en-US REACT=true NODE_ENV=production node --max_old_space_size=8192 node_modules/react-styleguidist/bin/styleguidist.js build --config scripts/styleguide.config.js --mode production",
```
[the remark this answers is not in the document]
Comment (jpan-box) [reply package.json:53]: All commands which eventually invoke webpack would need the additional option. 

The way I decided on which commands to add the option to was by looking at what gets called in our build pipelines. `build:prod:examples` does not run without the option, but it wasn't clear when that would be run. Should we go ahead and add it anywhere other node vars are set (e.g. any command with NODE_ENV)? or just add it "as-needed"?

[the remark this answers is not in the document]
Comment (tjuanitas) [reply package.json:53]: I think it runs here: https://github.com/box/box-ui-elements/blob/master/scripts/cut_release.sh#L135 which happens when building a new release

[the remark this answers is not in the document]
Comment (greg-in-a-box) [reply package.json:53]: build:prod:examples -> build -> build:dev:dist, so we dont need to add it to build:prod:examples

[the remark this answers is not in the document]
Comment (tjuanitas) [reply package.json:53]: isn't it calling the `build` from `styleguidist.js` and not the `build` script from package.json?

[the remark this answers is not in the document]
Comment (greg-in-a-box) [reply package.json:53]: hmmms good point

**Pull request closed without merging**